railgun/test/integration/udpingress_test.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/kong/kubernetes-ingress-controller/railgun/pkg/clientset"
+	"github.com/stretchr/testify/assert"
+)
+
+const dnsTestService = "quad9-dns"
+
+func TestMinimalUDPIngress(t *testing.T) {
+	if useLegacyKIC() {
+		t.Skip("legacy KIC does not support UDPIngress, skipping")
+	}
+	ctx := context.Background()
+
+	// gather the proxy container as it will need to be specially configured to serve UDP
+	proxy, err := cluster.Client().AppsV1().Deployments(controllers.DefaultNamespace).Get(ctx, "ingress-controller-kong", metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.Len(t, proxy.Spec.Template.Spec.Containers, 1)
+	container := proxy.Spec.Template.Spec.Containers[0].DeepCopy()
+
+	// override the KONG_STREAM_LISTEN env var in the proxy container
+	originalVal, err := overrideEnvVar(container, "KONG_STREAM_LISTEN", "0.0.0.0:9999 udp reuseport")
+	assert.NoError(t, err)
+	proxy.Spec.Template.Spec.Containers[0] = *container
+
+	// add the UDP port to the pod
+	container.Ports = append(container.Ports, corev1.ContainerPort{
+		Name:          dnsTestService,
+		ContainerPort: 9999,
+		Protocol:      corev1.ProtocolUDP,
+	})
+
+	// update the deployment with the new container configurations
+	proxy, err = cluster.Client().AppsV1().Deployments(controllers.DefaultNamespace).Update(ctx, proxy, metav1.UpdateOptions{})
+	assert.NoError(t, err)
+
+	// make sure we clean up after ourselves
+	defer func() {
+		// retrieve the current proxy
+		proxy, err := cluster.Client().AppsV1().Deployments(controllers.DefaultNamespace).Get(ctx, "ingress-controller-kong", metav1.GetOptions{})
+		assert.NoError(t, err)
+		container := proxy.Spec.Template.Spec.Containers[0].DeepCopy()
+		_, err = overrideEnvVar(container, "KONG_STREAM_LISTEN", originalVal.Value)
+		assert.NoError(t, err)
+
+		// remove the added UDP port
+		newPorts := make([]corev1.ContainerPort, 0, len(container.Ports)-1)
+		for _, port := range container.Ports {
+			if port.Name != dnsTestService {
+				newPorts = append(newPorts, port)
+			}
+		}
+		container.Ports = newPorts
+
+		// revert to pre-test state
+		proxy.Spec.Template.Spec.Containers[0] = *container
+		_, err = cluster.Client().AppsV1().Deployments(controllers.DefaultNamespace).Update(ctx, proxy, metav1.UpdateOptions{})
+		assert.NoError(t, err)
+
+		// ensure that the proxy deployment is ready before we proceed
+		assert.Eventually(t, func() bool {
+			d, err := cluster.Client().AppsV1().Deployments(controllers.DefaultNamespace).Get(ctx, proxy.Name, metav1.GetOptions{})
+			if err != nil {
+				t.Logf("WARNING: error while waiting for deployment %s to become ready: %v", proxy, err)
+				return false
+			}
+			if d.Status.ReadyReplicas == d.Status.Replicas && d.Status.AvailableReplicas == d.Status.Replicas && d.Status.UnavailableReplicas < 1 {
+				return true
+			}
+			return false
+		}, proxyUpdateWait, waitTick)
+	}()
+
+	// ensure that the proxy deployment is ready before we proceed
```

Comment:
For context, I found a related issue while testing this and reported it as https://github.com/Kong/kubernetes-ingress-controller/issues/1142